src/Microsoft.AspNetCore.OData/EnableNestedPathsAttribute.cs

```diff
@@ -0,0 +1,150 @@
+        }
+
+        /// <summary>
+        /// Transforms the result of the action based on the sequence of property accesses in the odata path
+        /// after the action is executed.
+        /// It first tries to retrieve the IQueryable from the
+        /// returning response message. It then uses the <see cref="ODataPathQueryBuilder"/> to transform
+        /// the query and sets the result back to the response message.
+        /// </summary>
+        /// <param name="actionExecutedContext">The context related to this action, including the response message,
+        /// request message and HttpConfiguration etc.</param>
+        public override void OnActionExecuted(ActionExecutedContext actionExecutedContext)
+        {
+            if (actionExecutedContext == null)
+            {
+                throw Error.ArgumentNull("actionExecutedContext");
+            }
+
+            HttpRequest request = actionExecutedContext.HttpContext.Request;
+            if (request == null)
+            {
+                throw Error.Argument("actionExecutedContext", SRResources.ActionExecutedContextMustHaveRequest);
+            }
+
+            ControllerActionDescriptor actionDescriptor = actionExecutedContext.ActionDescriptor as ControllerActionDescriptor;
+            if (actionDescriptor == null)
+            {
+                throw Error.Argument("actionExecutedContext", SRResources.ActionContextMustHaveDescriptor);
+            }
+
+            HttpResponse response = actionExecutedContext.HttpContext.Response;
+
+            ObjectResult responseContent = actionExecutedContext.Result as ObjectResult;
+            IQueryable result = responseContent.Value as IQueryable;
+            SingleResult singleResult = responseContent.Value as SingleResult;
+            if (singleResult != null)
+            {
+                // This could be a SingleResult, which has the property Queryable.
+                // But it could be a SingleResult() or SingleResult<T>. Sort by number of parameters
+                // on the property and get the one with the most parameters.
+                PropertyInfo propInfo = responseContent.Value.GetType().GetProperties()
+                    .OrderBy(p => p.GetIndexParameters().Count())
+                    .Where(p => p.Name.Equals("Queryable"))
+                    .LastOrDefault();
+
+                result = propInfo.GetValue(singleResult) as IQueryable;
+            }
+
+            if (result != null)
+            {
+                ODataPath path = actionExecutedContext.HttpContext.ODataFeature().Path;
+                IEdmModel model = actionExecutedContext.HttpContext.Request.GetModel();
+
+                var queryBuilder = new ODataPathQueryBuilder(result, model, path);
+                ODataPathQueryResult transformedResult = queryBuilder.BuildQuery();
+
+                if (transformedResult == null)
+                {
+                    // TODO: is this the best way to return 404
```

Comment:
TODO

Reply:
Plus I'm not sure how `EnableQuery` (if it also exists on the action method) will handle the `Result` if the value is `NotFoundObjectResult`...

Reply:
EnableQuery checks whether the result is successful before processing. If we set the result to a 404, then the user will get a 404 response. What I wasn't sure is whether it's okay to just create the not found result without passing an exception or error message to it. I don't know if there's any appropriate exception to pass at this point other than to just indicate that the resource does not exist.

Reply:
Consider the following current behaviors:
- `/Sales(13)` - You'll get a 404 if a resource with key 13 does not exist
- `/Sales?$filter=Id eq 13` - You'll get an empty array if a resource with Id equal to 13 does not exist
Since nested paths feature key segments, a 404 would appear like the logic result to return in this scenario.

However, consider this nested path scenario also
`/Sales(1)/Remark`
Supposing in whatever data source I'm using the value for `Remark` is `null`. Is your logic treating that as a value or responding with a `NotFound`?

Reply:
What would happen if a collection with mutliple nulls is encountered? would it filter out the nulls? return them? or simply end up having a 404 bubble up?

Reply:
@marabooy since you're accessing resources by their ids or properties, I'm struggling to picture the scenario you're laying out.

Suppose `Customer(1)/Orders` returns a collection. If the collection has null items, the response here should still be an array. If we want to access an entity in that collection, we'd access it by ID: `Customer(1)/Orders(2)` and in this case it doesn't matter what other items are null, we only care about the `Order` with id 2. If that doesn't exist, we return 404.

Reply:
The tricky part here is that there's currently no consistency in how nulls and 404 are handled. I've explained that in detail in this issue: https://github.com/OData/WebApi/issues/2479

My view is that we should address the null/404 issue in a separate PR because it seems to be an issue that already exists outside of this PR, rather than trying to handle it here.